kairos-prover/kairos-prover-risc0-server/src/main.rs

```diff
@@ -0,0 +1,210 @@
+use std::time::Instant;
+
+use axum::{http::StatusCode, Json};
+use axum_extra::routing::{RouterExt, TypedPath};
+use kairos_circuit_logic::{ProofInputs, ProofOutputs};
+
+// These constants represent the RISC-V ELF and the image ID generated by risc0-build.
+// The ELF is used for proving and the ID is used for verification.
+use methods::{PROVE_BATCH_ELF, PROVE_BATCH_ID};
+use risc0_zkvm::{ExecutorEnv, Receipt};
+use tracing::level_filters::LevelFilter;
+use tracing_subscriber::EnvFilter;
+
+#[tokio::main]
+async fn main() {
+    let _ = dotenvy::dotenv();
+    // Initialize tracing. In order to view logs, run `RUST_LOG=info cargo run`
+    tracing_subscriber::fmt()
+        .with_env_filter(tracing_subscriber::filter::EnvFilter::from_default_env())
+        .init();
+
+    let socket_addr = std::env::var("KAIROS_PROVER_SERVER_SOCKET_ADDR")
+        .expect("Failed to fetch environment variable KAIROS_PROVER_SERVER_SOCKET_ADDR");
+    let socket_addr = socket_addr
+        .parse::<std::net::SocketAddr>()
+        .expect("Failed to parse KAIROS_SERVER_SOCKET_ADDR");
+
+    let app = axum::Router::new()
+        .typed_post(prove_batch_route)
+        .with_state(());
+
+    tracing::info!("starting http server on `{}`", socket_addr);
+    let listener = tokio::net::TcpListener::bind(socket_addr).await.unwrap();
+    tracing::info!("listening on `{}`", socket_addr);
+    axum::serve(listener, app).await.unwrap()
+}
+
+#[derive(TypedPath, Debug, Clone, Copy)]
+#[typed_path("/api/v1/prove/batch")]
+pub struct ProveBatch;
+
+pub async fn prove_batch_route(
+    _: ProveBatch,
+    proof_inputs: Json<ProofInputs>,
+) -> Result<Json<(ProofOutputs, Receipt)>, (StatusCode, String)> {
+    let proof = tokio::task::spawn_blocking(move || prove_execution(proof_inputs.0))
+        .await
+        .map_err(|e| {
+            let e = format!("Error while joining proving task: {e}");
+            tracing::error!(e);
+            (StatusCode::INTERNAL_SERVER_ERROR, e)
+        })?;
+
+    let proof = proof.map_err(|e| {
+        let e = format!("Error while proving batch: {e}");
+        tracing::error!(e);
+        (StatusCode::INTERNAL_SERVER_ERROR, e)
+    })?;
+
+    Ok(Json(proof))
+}
+
+fn prove_execution(
+    proof_inputs: kairos_circuit_logic::ProofInputs,
+) -> Result<(ProofOutputs, Receipt), String> {
+    let timestamp = Instant::now();
+
+    let env = ExecutorEnv::builder()
+        .write(&proof_inputs)
+        .map_err(|e| format!("Error in ExecutorEnv builder write: {e}"))?
+        .build()
+        .map_err(|e| format!("Error in ExecutorEnv builder build: {e}"))?;
+
+    let receipt = risc0_zkvm::default_prover()
```

Comment:
[Default prover](https://docs.rs/risc0-zkvm/latest/risc0_zkvm/fn.default_prover.html) for non-dev environment is Bonsai. I think we should prefer generating proofs locally, either with `r0vm` subprocess or in-process.

**Request:** Please use explicit type of prover.

Reply:
For the demo it's not that important, it's not hard to fix though, I tend to agree that proving locally is generally better.

Reply:
Fixed
We were using local provers because BONSAI_API_URL and BONSAI_API_KEY were not set, but I agree picking the prover explicitly is better.

Reply:
I misinterpreted docs, Bonsai was not used then. Thanks for fix!

Reply:
Fixed in https://github.com/cspr-rad/kairos/pull/98/commits/ee016937731f06a4a1312c8e8dc399e2ee04d8ea.